src/client/address_book.cc

```diff
@@ -55,221 +55,292 @@
-AddressBook::~AddressBook() {
-  Stop();
+void AddressBookSubscriber::DownloadSubscriptionImpl() {
+  LogPrint(eLogInfo,
+      "AddressBookSubscriber: downloading subscription ", m_HTTP.GetURI().string(),
+      " ETag: ", m_HTTP.GetPreviousETag(),
+      " Last-Modified: ", m_HTTP.GetPreviousLastModified());
+  bool download_result = m_HTTP.Download();
+  if (download_result) {
+    std::stringstream stream(m_HTTP.GetDownloadedContents());
+    if (!m_Book.ValidateSubscriptionThenSaveToStorage(stream)) {
+      // Error during validation or storage, download again later
+      download_result = false;
+    }
+  }
+  m_Book.HostsDownloadComplete(download_result);
 }
 
-void AddressBook::Start(
-    std::shared_ptr<ClientDestination> local_destination) {
-  m_SharedLocalDestination = local_destination;
-  StartSubscriptions();
+void AddressBook::HostsDownloadComplete(
+    bool success) {
+  m_SubscriberIsDownloading = false;
+  if (m_SubscriberUpdateTimer) {
+    m_SubscriberUpdateTimer->expires_from_now(
+        boost::posix_time::minutes(
+            success ?
+            static_cast<std::uint16_t>(SubscriberTimeout::ContinuousUpdate) :
+            static_cast<std::uint16_t>(SubscriberTimeout::ContinuousRetry)));
+    m_SubscriberUpdateTimer->async_wait(
+        std::bind(
+            &AddressBook::SubscriberUpdateTimer,
+            this,
+            std::placeholders::_1));
+  }
 }
 
-void AddressBook::Stop() {
-  StopSubscriptions();
-  if (m_SubscriptionsUpdateTimer) {
-    m_SubscriptionsUpdateTimer.reset(nullptr);
+// TODO(unassigned): extend this to append new hosts (when other subscriptions are used)
+bool AddressBook::ValidateSubscriptionThenSaveToStorage(
+    std::istream& stream) {
+  // Save to subscription file if file does not exist or we have fresh download
+  std::unique_lock<std::mutex> lock(m_AddressBookMutex);
+  std::ofstream file;
+  if (!m_SubscriptionFileIsReady) {
+    file.open(
+        i2p::util::filesystem::GetFullPath(GetDefaultSubscriptionFilename()),
+        std::ofstream::out);
+    m_SubscriptionFileIsReady = true;
   }
-  if (m_IsDownloading) {
-    LogPrint(eLogInfo,
-        "AddressBook: subscription is downloading, waiting for termination");
-    for (int i = 0; i < 30; i++) {
-      if (!m_IsDownloading) {
-        LogPrint(eLogInfo, "AddressBook: subscription download complete");
-        break;
+  // Host per line and total number of host addresses
+  std::string host;
+  std::size_t num_addresses = 0;
+  // Read through stream, add to address book
+  while (!stream.eof()) {
+    getline(stream, host);
+    if (!host.length())
+      continue;  // skip empty line
+    // File ready, write/overwrite to disk
+    if (m_SubscriptionFileIsReady)
+      file << host << std::endl;
+    // Parse host from address, save to address book
+    std::size_t pos = host.find('=');
+    if (pos != std::string::npos) {
+      std::string name = host.substr(0, pos++);
+      std::string addr = host.substr(pos);
```

Comment:
@anonimal I'd make these `std::string` objects `const` if possible (but that's just me really).